### PR TITLE
Fold sign_holder_commitment_htlc_transactions into sign_holder_commitment

### PR DESCRIPTION
**Pull request by devrandom**
Signing the commitment transaction is almost always followed by signing the attached HTLC transactions, so fold the signing operations into a single method.

Followup to #742.

### CODE REVIEW

**Comment by devrandom** [comment]
I notice that `get_fully_signed_htlc_tx` and `get_fully_signed_holder_tx` do nothing if there's a failure to sign.  It seems like these should panic.

**Comment by devrandom** [comment]
Made all signing failures panic in `OnChainTx`.  Such failures can result in losing funds, so should not happen.

File: lightning/src/ln/onchaintx.rs [hunk 1]
```diff
@@ -939,23 +925,21 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
...
-			}
+			let (sig, htlc_sigs) = self.key_storage.sign_holder_commitment(holder_commitment, &self.secp_ctx).expect("sign holder commitment");
+			self.holder_htlc_sigs = Some(Self::extract_holder_sigs(holder_commitment, htlc_sigs));
+			Some(holder_commitment.add_holder_sig(funding_redeemscript, sig))
 		} else {
 			None
 		}
 	}
 
+	// The current holder commitment tx MUST have been signed before calling this with one of its outpoints.
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs:936]: I don't see how we can guarantee this without changes in ChannelMonitor - you can have multiple ChannelMonitors, one which signed and broadcasted a commitment tx, and another which is now trying to broadcast HTLC transactions without ever having signed the commitment transaction(s).

**Review by TheBlueMatt — COMMENTED** [review]
CC @ariard.

**Comment by devrandom** [comment]
Given [this comment](https://github.com/rust-bitcoin/rust-lightning/pull/770#discussion_r552788460), I'm thinking that there's not enough motivation for this change and it's OK for the commitment Tx signing to continue to be decoupled from the HTLC signing.

Should I close this and open a different PR for the second commit?

**Comment by TheBlueMatt** [comment]
> Given this comment, I'm thinking that there's not enough motivation for this change and it's OK for the commitment Tx signing to continue to be decoupled from the HTLC signing.

I was/am OK with it given the API simplification, which in the vast-majority case folds two back-to-back calls into one, but I certainly don't feel strong about it - up to you if you want to pursue it.

> Should I close this and open a different PR for the second commit?

Sure, the fact that a user-returned Err from ChanSigner implies you can never claim your funds needs to be documented, at least, and panic makes sense.

File: lightning/src/ln/onchaintx.rs [hunk 2]
```diff
@@ -925,12 +914,9 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
...
-				Ok(sig) => {
-					Some(holder_commitment.add_holder_sig(funding_redeemscript, sig))
-				},
-				Err(_) => return None,
-			}
+			let (sig, htlc_sigs) = self.key_storage.sign_holder_commitment(holder_commitment, &self.secp_ctx).expect("signing holder commitment");
+			self.holder_htlc_sigs = Some(Self::extract_holder_sigs(holder_commitment, htlc_sigs));
+			Some(holder_commitment.add_holder_sig(funding_redeemscript, sig))
 		} else {
 			None
```
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs:921]: The function comment indicates that `self.holder_commitment` can be validly `None`.  Do you think the comment is wrong?

File: lightning/src/ln/onchaintx.rs (hunk 1 above)
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs]: I reworked this to remove this assumption, and also added relevant docs to explain why we might still need to sign the HTLCs separately in some cases.

**Comment by devrandom** [comment]
OK, I'm going to give this a try and we can see if we like the result.  I pushed code changes that address that comment, but I still need to address a couple of other things.

File: lightning/src/chain/keysinterface.rs [hunk 3]
```diff
@@ -239,7 +239,7 @@ pub trait ChannelKeys : Send+Clone {
 	/// An external signer implementation should check that the commitment has not been revoked.
 	//
 	// TODO: Document the things someone using this interface should enforce before signing.
-	fn sign_holder_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, commitment_tx: &HolderCommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<Signature, ()>;
+	fn sign_holder_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, commitment_tx: &HolderCommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<(Signature, Vec<Signature>), ()>;
```
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:242]: Docs completely revised.

File: lightning/src/ln/onchaintx.rs [hunk 4]
```diff
@@ -594,17 +594,15 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
...
-								bumped_tx.input[i].witness[0].push(SigHashType::All as u8);
-								if htlc.is_some() {
-									bumped_tx.input[i].witness.push(chan_keys.revocation_key.clone().serialize().to_vec());
-								} else {
-									bumped_tx.input[i].witness.push(vec!(1));
-								}
-								bumped_tx.input[i].witness.push(witness_script.clone().into_bytes());
-							} else { return None; }
-							//TODO: panic ?
+							let sig= self.key_storage.sign_justice_transaction(&bumped_tx, i, *amount, &per_commitment_key, htlc, &self.secp_ctx).expect("sign justice tx");
```
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs]: I assume you meant docs on `generate_claim_tx` - this is done.

File: lightning/src/chain/keysinterface.rs [hunk 5]
```diff
@@ -233,13 +233,16 @@ pub trait ChannelKeys : Send+Clone {
 	// TODO: Document the things someone using this interface should enforce before signing.
 	fn sign_counterparty_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, commitment_tx: &CommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<(Signature, Vec<Signature>), ()>;
 
-	/// Create a signature for a holder's commitment transaction. This will only ever be called with
-	/// the same commitment_tx (or a copy thereof), though there are currently no guarantees
-	/// that it will not be called multiple times.
+	/// Create a signatures for a holder's commitment transaction and its claiming HTLC transactions.
+	/// This will only ever be called with a non-revoked commitment_tx.  This may be the latest
```
Comment (ariard) [inline lightning/src/chain/keysinterface.rs:237]: I think we can be better on this comment.

"A signing entity (watchtower/ChannelMonitor replica) will only _initiate_ signing of a commitment tx for the latest local state. Due to a distributed setup, it might have to _react_ to a commitment tx broadcast and sign HTLC for the previous-latest local state." We should distinguish signing initiation from signing reaction.

W.r.t to "secondary ChannelMonitor", it designates one replica of a distributed ChannelMonitor setup, we should bind to one nomenclature and always call them "ChannelMonitor replica", as you might not have a "primary" at all in that kind of setup. 

If your HTLC state is empty, your Vec<Signature> might be empty, and that's completely fine.

"This may be called multiple times", I think the design purpose of this interface liberality is to handle reorg or concurrent broadcast, do we have other cases ?

File: lightning/src/ln/onchaintx.rs [hunk 6]
```diff
@@ -474,6 +474,8 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
 
 	/// Lightning security model (i.e being able to redeem/timeout HTLC or penalize coutnerparty onchain) lays on the assumption of claim transactions getting confirmed before timelock expiration
 	/// (CSV or CLTV following cases). In case of high-fee spikes, claim tx may stuck in the mempool, so you need to bump its feerate quickly using Replace-By-Fee or Child-Pay-For-Parent.
+	/// Panics if there are signing errors, because signing operations in reaction to on-chain events
+	/// are not expected to fail, and if they do, we may lose funds.
```
Comment (ariard) [inline lightning/src/ln/onchaintx.rs:478]: Depends, I think this comment is excluding hardware/system failures, which make sense, but what about more subtle cases in the future, like a BOLT13 watchtower client providing garbage for an obfuscated justice transaction ?

File: lightning/src/ln/onchaintx.rs [hunk 7]
```diff
@@ -890,20 +892,29 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
...
+		if self.holder_htlc_sigs.is_none() {
+			if let Some(ref holder_commitment) = self.holder_commitment {
+				let (_sig, sigs) = self.key_storage.sign_holder_commitment(holder_commitment, &self.secp_ctx).expect("sign holder commitment");
 				self.holder_htlc_sigs = Some(Self::extract_holder_sigs(holder_commitment, sigs));
 			}
 		}
 	}
 
+	// Normally only the latest commitment tx and HTLCs need to be signed.  However, in some
+	// configurations we may have updated our holder commtiment but a replica of the ChannelMonitor
```
Comment (ariard) [inline lightning/src/ln/onchaintx.rs:909]: *commitment

File: lightning/src/ln/onchaintx.rs [hunk 8]
```diff
@@ -966,7 +979,7 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
 				}
 			}
 		}
-		if self.prev_holder_commitment.is_some() {
+		if htlc_tx.is_none() && self.prev_holder_commitment.is_some() {
```
Comment (ariard) [inline lightning/src/ln/onchaintx.rs:971]: I don't think we have test coverage for this code path ?

**Review by ariard — COMMENTED** [review]
I'm Concept ACK on this, it's clearly a simplification of those critical on-chain code paths.

File: lightning/src/chain/keysinterface.rs [hunk 9]
```diff
@@ -233,13 +233,16 @@ pub trait ChannelKeys : Send+Clone {
...
+	/// tx or it may be the second latest if not revoked and some watchtower/secondary
+	/// ChannelMonitor decided to broadcast before it had been updated to the latest.
+	/// This may be called multiple times.
```
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs:240]: nit: maybe mention that it can be called multiple times for the same commitment_tx - its obviously called regularly for different ones.

File: lightning/src/ln/onchaintx.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/onchaintx.rs:940]: Hmm, oops, but also isn't that out of date since 5d0bfa383425ecbf308fdf819dbf7ad19e6c5ca4? The comment describes the time between `get_outbound_funding_created` and `funding_signed` and since 80055d4bb41b83428f74141c0140165add2609a7 we remove some of the `Option`s which can cause it.

File: lightning/src/ln/onchaintx.rs (hunk 6 above)
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs:478]: Wouldn't that be a different code path?  (and does it exist yet?)

File: lightning/src/ln/onchaintx.rs (hunk 8 above)
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs:971]: it's covered by `claim_htlc_outputs.*` and some other functional tests.

File: lightning/src/ln/onchaintx.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs]: nice catch!  I de-optionated things in a new commit.

File: lightning/src/chain/keysinterface.rs (hunk 9 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:240]: > called multiple times for the same commitment_tx - its obviously called regularly for different ones.

I agree with the first part and will adjust the doc, but I don't understand what you mean by the second part.  It is *not* called regularly - it's only called at force-close which happens rarely, right?

File: lightning/src/chain/keysinterface.rs (hunk 5 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: reworded, see if you like it.

yes, I think the multiple-times applies to reorgs, but it seems to me that fee bumps also cause re-signing if I understand the `OnChainTxHandler` code

Comment (ariard) [reply lightning/src/chain/keysinterface.rs]: See the comment "Timer set to $NEVER given we can't bump tx without anchor outputs". This code path should execute only once, pre-anchor we can't bump holder's {commitment, htlc} txn. But right won't stay true for long.

File: lightning/src/ln/onchaintx.rs (hunk 6 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs:495]: It doesn't exist yet but a while back I draw a schema for watchtower-support, see : https://github.com/rust-bitcoin/rust-lightning/issues/606#issuecomment-619338323. The first step is to land #642. But it will be the same code path, hitting `generate_claim_tx`. 

Nevermind, we'll update the comment when we have to.

File: lightning/src/ln/onchaintx.rs (hunk 8 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs:976]: Yes covered after tweaking it 

```
failures:
    ln::functional_tests::test_bump_penalty_txn_on_revoked_htlcs
    ln::functional_tests::test_justice_tx
    ln::functional_tests::test_static_spendable_outputs_justice_tx_revoked_htlc_success_tx
    ln::functional_tests::test_static_spendable_outputs_justice_tx_revoked_htlc_timeout_tx
```

File: lightning/src/chain/keysinterface.rs [hunk 10]
```diff
@@ -233,13 +233,18 @@ pub trait ChannelKeys : Send+Clone + Writeable {
 	// TODO: Document the things someone using this interface should enforce before signing.
 	fn sign_counterparty_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, commitment_tx: &CommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<(Signature, Vec<Signature>), ()>;
 
-	/// Create a signature for a holder's commitment transaction. This will only ever be called with
-	/// the same commitment_tx (or a copy thereof), though there are currently no guarantees
-	/// that it will not be called multiple times.
+	/// Create a signatures for a holder's commitment transaction and its claiming HTLC transactions.
+	/// This will only ever be called with a non-revoked commitment_tx.  This will be called with the
+	/// latest commitment_tx when we initiate a force-close.
+	/// This will be called with the second latest, just to get claiming HTLC signatures, if we are
```
Comment (ariard) [inline lightning/src/chain/keysinterface.rs:239]: Note, we used very consistently previous latest as a variable name or across comments. Maybe better to stick to it.

File: lightning/src/chain/keysinterface.rs [hunk 11]
```diff
@@ -233,13 +233,18 @@ pub trait ChannelKeys : Send+Clone + Writeable {
...
+	/// reacting to a watchtower/ChannelMonitor replica that decided to broadcast before it had been
```
Comment (ariard) [inline lightning/src/chain/keysinterface.rs:240]: According to our landing-soon GLOSSARY, a watchtower don't' have the capability to broadcast a commitment tx, so would remove it and only mention replica :) ?

File: lightning/src/chain/keysinterface.rs [hunk 12]
```diff
@@ -233,13 +233,18 @@ pub trait ChannelKeys : Send+Clone + Writeable {
...
+	/// updated to the latest.
+	/// This may be called multiple times for the same transaction.
+	///
 	/// An external signer implementation should check that the commitment has not been revoked.
 	//
 	// TODO: Document the things someone using this interface should enforce before signing.
-	fn sign_holder_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, commitment_tx: &HolderCommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<Signature, ()>;
+	fn sign_holder_commitment<T: secp256k1::Signing + secp256k1::Verification>(&self, commitment_tx: &HolderCommitmentTransaction, secp_ctx: &Secp256k1<T>) -> Result<(Signature, Vec<Signature>), ()>;
```
Comment (ariard) [inline lightning/src/chain/keysinterface.rs:247]: Should we change the name to `sign_holder_state/commitment_and_htlcs`, to convey that this method now scopes HTLC txn signature.

File: lightning/src/ln/onchaintx.rs [hunk 13]
```diff
@@ -906,20 +908,29 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
 
 	pub(crate) fn provide_latest_holder_tx(&mut self, tx: HolderCommitmentTransaction) {
 		self.prev_holder_commitment = self.holder_commitment.take();
+		self.holder_htlc_sigs = None;
 		self.holder_commitment = Some(tx);
 	}
 
+	// Normally holder HTLCs are signed at the same time as the holder commitment tx.  However,
+	// in some configurations, the holder commitment tx has been signed and broadcast by a secondary
+	// ChannelMonitor, so we handle that case here.
```
Comment (ariard) [inline lightning/src/ln/onchaintx.rs:914]: `ChannelMonitor` replica ?

File: lightning/src/ln/onchaintx.rs [hunk 14]
```diff
@@ -612,17 +612,15 @@ impl<ChanSigner: ChannelKeys> OnchainTxHandler<ChanSigner> {
...
-								bumped_tx.input[i].witness[0].push(SigHashType::All as u8);
-								if htlc.is_some() {
-									bumped_tx.input[i].witness.push(chan_keys.revocation_key.clone().serialize().to_vec());
-								} else {
-									bumped_tx.input[i].witness.push(vec!(1));
-								}
-								bumped_tx.input[i].witness.push(witness_script.clone().into_bytes());
-							} else { return None; }
-							//TODO: panic ?
+							let sig = self.key_storage.sign_justice_transaction(&bumped_tx, i, *amount, &per_commitment_key, htlc, &self.secp_ctx).expect("sign justice tx");
```
Comment (ariard) [inline lightning/src/ln/onchaintx.rs:616]: I do have a doubt on this change, I'm not sure an external signer failure should bring down your Lightning node. You might a partial disk failure, e.g corrupted key data for channel A, but still your signer might still succeed to produce signatures for channel B. 

Or this a kind of scenario which should be managed behind the key interface and a dummy sig should be returned ?

**Review by ariard — COMMENTED** [review]
I'm good with the code changes but see the comment about "Panic if signing fails in OnChainTx"

File: lightning/src/chain/keysinterface.rs (hunk 11 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:240]: oh, right

File: lightning/src/chain/keysinterface.rs (hunk 12 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:247]: agreed

File: lightning/src/ln/onchaintx.rs (hunk 13 above)
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs:914]: :+1:

File: lightning/src/ln/onchaintx.rs (hunk 14 above)
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs:616]: Here are my thoughts on this.

I believe that a highly available installation should have replication at these layers:

- signer - there should be a quorum of signers that have to agree for channel state advancement
- chanmon - there should be at least one chanmon replica reacting to on-chain events
- network handler (everything other than signer/chanmon) - probably want this in a active+hot-standby configuration, but maybe you can even do active+active or split by channel

I think in such a configuration, the signer layer should have a "TEMPFAIL" result, which should defer the operation and retry later.  For example, there may not be a quorum for one channel, but another channel can go ahead.  I think this requires some additional code to implement the deferral, so should be its own effort.  However, the existing errors this function produced are not in that category, they indicated a bug or corruption.

I think a faulty disk should probably result in the replica brought down and operator intervention, since it's hard to automatically figure out if the replica is still sane (maybe we are seeing old data, maybe our RAM is bad).  In a replicated environment, there should be enough other replicas to continue.

To summarize, I think the existing errors should be fatal, but in the future we should create a new kind of temporary error in all signer API calls.  This kind of error should be generated by the signer "client-side" if it fails to arrange for a quorum and should cause the operation to be deferred.

Comment (TheBlueMatt) [reply lightning/src/ln/onchaintx.rs:616]: > You might a partial disk failure, e.g corrupted key data for channel A, but still your signer might still succeed to produce signatures for channel B.

I'd strongly suggest that this is a *great* reason to bring down a lightning node - your hardware is corrupting data, for all we know the data may be corrupted in a way that results in us revealing a private key.

File: lightning/src/chain/keysinterface.rs (hunk 10 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: :+1:

File: lightning/src/ln/onchaintx.rs (hunk 14 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs:616]: I like the idea of temporary errors generated by the "client-side" signer, I agree that's a good direction for the future.

> your hardware is corrupting data, for all we know the data may be corrupted in a way that results in us revealing a private key.

It's hard to say, on one hand, if you have only one replica and you're crashing your only monitoring instance you might lose all your channels instead of burning only one. And the other hand, your hardware failure might be so gross that you reveal private key of L1 funds if your signer is cross-layer. 

I don't think there is _a_ good answer and it's likely function of your configuration, funds at stake, confidence in your hardware, ... One more trade-off we will have to document somewhere I guess.

Comment (TheBlueMatt) [reply lightning/src/ln/onchaintx.rs:616]: Sure, there isn't a clear answer, but we do know its critical that the user get involved ASAP and figure out what issue is happening. Continuing on is usually a great way to not have the user get involved, and I'd argue that the only real way we have of doing that is to panic.

**Comment by ariard** [comment]
Code Review ACK d657353, thanks to update with the latest suggestions.

File: lightning/src/ln/onchaintx.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/onchaintx.rs]: Oops, sorry I missed this, no, I meant on `sign_holder_commitment_and_htlcs` - `generate_claim_tx` isn't pub (so the docs here aren't visible), and the panic impacts all the way up to users (as it can be called in a number of contexts related to `ChannelMonitor`). Alternatively, we could note it in `ChannelMonitor` in all the `pub` functions that call through to here, but the call graph isn't exactly trivial.

**Review by TheBlueMatt — APPROVED** [review]
Aside from additional docs at https://github.com/rust-bitcoin/rust-lightning/pull/770#discussion_r559713370, LGTM.

File: lightning/src/ln/onchaintx.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs]: Added comment on `sign_holder_commitment_and_htlcs`

**Review by TheBlueMatt — APPROVED** [review]
